macros/contexts/contextLinearRelation.pl

```diff
@@ -0,0 +1,399 @@
+			formulaClass  => "LinearRelation"
+		},
+
+		'<=' => {
+			precedence    => .5,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' <= ',
+			TeX           => '\le ',
+			kind          => "le",
+			class         => 'LinearRelation::inequality',
+			formulaClass  => "LinearRelation",
+			alternatives  => ["\x{2264}"]
+		},
+		'=<' => {
+			precedence    => .5,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' <= ',
+			TeX           => '\le ',
+			kind          => "le",
+			class         => 'LinearRelation::inequality',
+			formulaClass  => "LinearRelation"
+		},
+
+		'>=' => {
+			precedence    => .5,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' >= ',
+			TeX           => '\ge ',
+			kind          => 'ge',
+			reverse       => 'le',
+			class         => 'LinearRelation::inequality',
+			formulaClass  => "LinearRelation",
+			alternatives  => ["\x{2265}"]
+		},
+		'=>' => {
+			precedence    => .5,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' >= ',
+			TeX           => '\ge ',
+			kind          => 'ge',
+			reverse       => 'le',
+			class         => 'LinearRelation::inequality',
+			formulaClass  => "LinearRelation"
+		},
+
+		'!=' => {
+			precedence    => .5,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' != ',
+			TeX           => '\ne ',
+			kind          => 'ne',
+			reverse       => 'ne',
+			class         => 'LinearRelation::inequality',
+			formulaClass  => "LinearRelation",
+			alternatives  => ["\x{2260}"]
+		},
+		'<>' => {
+			precedence    => .5,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' != ',
+			TeX           => '\ne ',
+			kind          => 'ne',
+			reverse       => 'ne',
+			class         => 'LinearRelation::inequality',
+			formulaClass  => "LinearRelation"
+		}
+	);
+	main::PG_restricted_eval('sub LinearRelation {Value->Package("LinearRelation()")->new(@_)}');
+}
+
+sub new {
+	my $self    = shift;
+	my $class   = ref($self) || $self;
+	my $context = (Value::isContext($_[0]) ? shift : $self->context);
+	my $formula = "Value::Formula";
+	return shift if scalar(@_) == 1 && ref($_[0]) eq 'LinearRelation';
+	$_[0] = $context->Package("Point")->new($context, $_[0])  if ref($_[0]) eq 'ARRAY';
+	$_[1] = $context->Package("Vector")->new($context, $_[1]) if ref($_[1]) eq 'ARRAY';
+
+	my ($p, $N, $plane, $vars, $d);
+	if (scalar(@_) >= 2 && Value::classMatch($_[0], 'Point', 'Vector') && Value::classMatch($_[1], 'Vector')
+		|| Value::isRealNumber($_[1]))
+	{
+		#
+		# Make a plane from a point and a vector and optionally a
+		# symbol <= or = or =>,
+		# e.g. LinearRelation($point, $vector, '<=');
+		# or from a list of coefficients and the constant
+		# e.g. LinearRelation($dist, [3,5,6], '<=')
+		# one can optionally add new Context
+		# variables ($point, $vector, '<=',[qw(a1,a2,a3)])
+		$p = shift;
```

Comment:
We're trying to get away from using shift to pull arguments lower in a subroutine.  
On line 203 or 204 ` my ($p, $N, $bop, $vars) = @_;`

Reply:
This is tricky because if the 2nd argument is a `Vector`, the arguments are interpreted one way. Otherwise, a different way. So what is `$N` in one conditional block is `$vars` in the other. It's like `my ($p, $N, $bop, $vars) = @_;` in one case and `my ($plane, $vars) = @_;` in the other. Since `$vars` makes sense as a name for the 4th argument sometimes but the 2nd argument other times, that throws a wrench in naming these all at the head of the subroutine.

Reply:
That is actually a problem with much of how PG is designed.  Its design has not followed good programming practices, and that has made things very confusing and unclear for future development.  This is a pain for those that are familiar with the PG design, and is extremely bad for those not already familiar with the design and trying to get started with it.

Optional arguments should always be at the end.  This is true in any language. It is unfortunate that MathObjects in general have the optional context argument as the first argument.

Reply:
>  Its design has not followed good programming practices

Of course what constitutes "good programming practices" changes as the language and the paradigms that it supports evolve.

> Optional arguments should always be at the end. This is true in any language.

A rather bold statement.  A number of languages allow alternative call signatures for a given function, particularly for an untyped language like Perl.  For example, in this case we have `LinearRelation(point, normal)` and `LinearRelation(formula)`, and the other MathObjects all work similarly (e.g., `Complex($a, $b)`, `Complex([$a, $b])`, `Complex("$a + $bi")`, and so on). So one could think of this as having two additional signatures, `LinearRelation(context, point, normal)` and `LinearRelation(context, formula)`.

This gets used to make MathObjects in a particular context that may not be the current context, and since it is used on generic objects, in order for the context to be at the end of the constructor argument list, MathObjects would need to know the argument lists of all possible MathObjects (an open-ended list), in order to be able to put it in the right position for each kind of MathObject that may need to be created.  Moreover, since some MathObjects have arbitrary length argument lists, like `Point($a, $b, $c, $d)`, there is no optional position at the end of the list, as the length of the list is not specified.  

You might suggest that one looks to see if the last argument is a context and remove that and use it if so.  But, some objects take key-value pairs as the final arguments (as Alex does here with the remaining arguments after `$p`, `$N`, `$bop`, and `$vars` being used as arguments to `reduce()`).  So in that case, the last "argument" could be a pair like `context => $context`, which would make the last two arguments in `@_` be `context` and $context.  One could not easily distinguish this from the last argument being an optional context argument.

While it would be possible to use

``` perl
my ($p, $N, $bop, $vars, %reduce) = @_;
```

to process the arguments (once the context has been shifted off), it would *not* work to do 

``` perl
my ($p, $N, $bop, $vars, %reduce, $context) = @_;
```

to capture the context.  You would have to do processing to remove the context from the argument list whether it is at the end or at the beginning, so I don't think there is anything to be gained by moving it to the end, and in fact some potential complications when doing so.

Perhaps thinking about this as the constructor having additional call signatures will help it make more sense to you.

Reply:
"Perl Best Practices" (implemented by perlcritic at severity level 4) recommends unpacking @_ on the first line of the function to achieve this clarity of what arguments are expected.

The point is to make the arguments for the method clear so that developers know by looking at line 1 of the method (or in the sub definition itself if signatures were enabled) what arguments are expected.  Different call signatures (in C or C++ for instance) does this.  You can look at the function headers and you know what is expected (for C and C++ even types are precise).  Perl doesn't have the concept of alternate call signatures in such a clear and concise way.

I am not saying that the MathObject code should now be rewritten to change the way the optional context argument is written at this point, and in some cases achieving what would be done in other languages with call signatures is not entirely possible in a way that will satisfy perlcritic.  This is where documentation comes in.  Comments or POD before the function would help to clarify these things.